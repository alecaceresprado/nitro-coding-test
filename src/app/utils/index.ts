export * from './post.grouper.utils';
