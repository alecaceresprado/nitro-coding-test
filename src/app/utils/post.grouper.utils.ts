import { Post, OrderCriteria } from '@models';

export const groupBy = (posts: Post[], criteria: OrderCriteria) =>
  posts.reduce((acc, curr) => {
    let parent;
    switch (criteria) {
      case OrderCriteria.WEEK:
        const date = new Date(Number(curr.time));
        parent = `Week ${getWeekNumber(date)}`;
        break;
      case OrderCriteria.LOCATION:
        parent = curr.location;
        break;
      case OrderCriteria.AUTHOR:
        parent = curr.author;
        break;
    }
    acc[parent] = [...(acc[parent] || []), curr];
    return acc;
  }, {});

function getWeekNumber(date: Date): number {
  const firstDayOfYear: Date = new Date(date.getFullYear(), 0, 1);
  const pastDaysOfYear: number =
    (date.getTime() - firstDayOfYear.getTime()) / 86400000;
  return Math.ceil((pastDaysOfYear + firstDayOfYear.getDay() + 1) / 7);
}
