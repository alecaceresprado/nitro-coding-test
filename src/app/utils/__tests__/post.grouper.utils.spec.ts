import { OrderCriteria } from '@models';
import { groupBy } from '../post.grouper.utils';
const mockDb = require('../../../../db.json');

describe('groupBy', () => {
  it('should return empty array', () => {
    expect(groupBy([], OrderCriteria.WEEK)).toMatchSnapshot();
  });

  it('should return weekGroup', () => {
    expect(groupBy(mockDb.posts, OrderCriteria.WEEK)).toMatchSnapshot();
  });

  it('should return location', () => {
    expect(groupBy(mockDb.posts, OrderCriteria.LOCATION)).toMatchSnapshot();
  });

  it('should return author', () => {
    expect(groupBy(mockDb.posts, OrderCriteria.AUTHOR)).toMatchSnapshot();
  });
});
