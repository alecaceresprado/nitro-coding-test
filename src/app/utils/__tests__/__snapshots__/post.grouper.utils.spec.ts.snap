// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupBy should return author 1`] = `
Object {
  "Happy Developer": Array [
    Object {
      "author": "Happy Developer",
      "id": 3,
      "location": "SanFrancisco",
      "text": "Digital transformation isn’t justa buzzword",
      "time": "1501690872999",
    },
    Object {
      "author": "Happy Developer",
      "id": 4,
      "location": "Sydney",
      "text": "An expectation of digital efficiency has become the norm in our daily lives",
      "time": "1591699872999",
    },
  ],
  "Happy Manager": Array [
    Object {
      "author": "Happy Manager",
      "id": 5,
      "location": "Dublin",
      "text": "A modern PDF annotator that canaccommodate all of the cooks in a very busy kitchen is what your employees really need.",
      "time": "1553080742",
    },
    Object {
      "author": "Happy Manager",
      "id": 6,
      "location": "Dublin",
      "text": "An integratedproductivity solution breaks information through barriers and allows workers to collaborate in real time.",
      "time": "1553099742",
    },
  ],
  "Happy User": Array [
    Object {
      "author": "Happy User",
      "id": 1,
      "location": "San Francisco",
      "text": "Proper PDF conversion ensures that every element of your document remains just as you left it.",
      "time": "1591699872486",
    },
    Object {
      "author": "Happy User",
      "id": 2,
      "location": "San Francisco",
      "text": "The modernworkplace is increasingly digital, and workflows are constantly evolving. ",
      "time": "1501690872486",
    },
  ],
}
`;

exports[`groupBy should return empty array 1`] = `Object {}`;

exports[`groupBy should return location 1`] = `
Object {
  "Dublin": Array [
    Object {
      "author": "Happy Manager",
      "id": 5,
      "location": "Dublin",
      "text": "A modern PDF annotator that canaccommodate all of the cooks in a very busy kitchen is what your employees really need.",
      "time": "1553080742",
    },
    Object {
      "author": "Happy Manager",
      "id": 6,
      "location": "Dublin",
      "text": "An integratedproductivity solution breaks information through barriers and allows workers to collaborate in real time.",
      "time": "1553099742",
    },
  ],
  "San Francisco": Array [
    Object {
      "author": "Happy User",
      "id": 1,
      "location": "San Francisco",
      "text": "Proper PDF conversion ensures that every element of your document remains just as you left it.",
      "time": "1591699872486",
    },
    Object {
      "author": "Happy User",
      "id": 2,
      "location": "San Francisco",
      "text": "The modernworkplace is increasingly digital, and workflows are constantly evolving. ",
      "time": "1501690872486",
    },
  ],
  "SanFrancisco": Array [
    Object {
      "author": "Happy Developer",
      "id": 3,
      "location": "SanFrancisco",
      "text": "Digital transformation isn’t justa buzzword",
      "time": "1501690872999",
    },
  ],
  "Sydney": Array [
    Object {
      "author": "Happy Developer",
      "id": 4,
      "location": "Sydney",
      "text": "An expectation of digital efficiency has become the norm in our daily lives",
      "time": "1591699872999",
    },
  ],
}
`;

exports[`groupBy should return weekGroup 1`] = `
Object {
  "Week 24": Array [
    Object {
      "author": "Happy User",
      "id": 1,
      "location": "San Francisco",
      "text": "Proper PDF conversion ensures that every element of your document remains just as you left it.",
      "time": "1591699872486",
    },
    Object {
      "author": "Happy Developer",
      "id": 4,
      "location": "Sydney",
      "text": "An expectation of digital efficiency has become the norm in our daily lives",
      "time": "1591699872999",
    },
  ],
  "Week 31": Array [
    Object {
      "author": "Happy User",
      "id": 2,
      "location": "San Francisco",
      "text": "The modernworkplace is increasingly digital, and workflows are constantly evolving. ",
      "time": "1501690872486",
    },
    Object {
      "author": "Happy Developer",
      "id": 3,
      "location": "SanFrancisco",
      "text": "Digital transformation isn’t justa buzzword",
      "time": "1501690872999",
    },
  ],
  "Week 4": Array [
    Object {
      "author": "Happy Manager",
      "id": 5,
      "location": "Dublin",
      "text": "A modern PDF annotator that canaccommodate all of the cooks in a very busy kitchen is what your employees really need.",
      "time": "1553080742",
    },
    Object {
      "author": "Happy Manager",
      "id": 6,
      "location": "Dublin",
      "text": "An integratedproductivity solution breaks information through barriers and allows workers to collaborate in real time.",
      "time": "1553099742",
    },
  ],
}
`;
